04.ls/ls.rb

```diff
@@ -33,4 +59,93 @@ def split_files(files)
   divided_files
 end
 
+def show_long_format(path, file_names)
+  stats_by_path = file_names.to_h do |file_name|
+    absolute_path = File.join(path, file_name)
+    [absolute_path, File.lstat(absolute_path)]
+  end
+  max_lengths = result_max_lengths(stats_by_path.values)
+
+  puts "total #{stats_by_path.values.sum(&:blocks)}"
+  generate_file_data_list(stats_by_path).each do |file_data|
+    puts format_file(file_data, max_lengths)
+  end
+end
+
+def generate_file_data_list(stats_by_path)
+  stats_by_path.map do |absolute_path, stat|
+    {
+      permission: to_permission(stat),
+      link_count: stat.nlink.to_s,
+      owner: Etc.getpwuid(stat.uid).name,
+      group: Etc.getgrgid(stat.gid).name,
+      byte_size: stat.size.to_s,
+      time_stamp: stat.mtime.strftime(to_timestamp(stat.mtime)),
+      file_name: to_file_name(stat, absolute_path)
+    }
+  end
+end
+
+def to_timestamp(mtime)
+  (Time.now - mtime) >= HALF_YEAR ? ' %_m %_d  %Y ' : ' %_m %_d %H:%M '
+end
+
+def to_file_name(stat, absolute_path)
+  file_name = File.basename(absolute_path)
+  stat.symlink? ? "#{file_name} -> #{File.readlink(absolute_path)}" : file_name
+end
+
+def format_file(file_data, max_lengths)
+  space = ' '
+  [
+    file_data[:permission], space * 2,
+    file_data[:link_count].rjust(max_lengths[:link]), space,
+    file_data[:owner].ljust(max_lengths[:owner]), space * 2,
+    file_data[:group].ljust(max_lengths[:group]), space * 2,
+    file_data[:byte_size].rjust(max_lengths[:byte_size]),
+    file_data[:time_stamp],
+    file_data[:file_name]
+  ].join
+end
+
+def result_max_lengths(stat_files)
+  lengths = { link: 0, owner: 0, group: 0, byte_size: 0 }
+
+  stat_files.each do |stat|
+    lengths[:link] = stat.nlink.to_s.length if lengths[:link] < stat.nlink.to_s.length
+    lengths[:owner] = Etc.getpwuid(stat.uid).name.length if lengths[:owner] < Etc.getpwuid(stat.uid).name.length
+    lengths[:group] = Etc.getgrgid(stat.gid).name.length if lengths[:group] < Etc.getgrgid(stat.gid).name.length
+    lengths[:byte_size] = stat.size.to_s.length if lengths[:byte_size] < stat.size.to_s.length
+  end
+  lengths
+end
+
+def to_permission(stat)
+  to_characters = divide_characters(stat).map do |section, char|
+    to_special_permission(stat, section, char) || PERMISSIONS[char]
+  end
+  FILE_TYPES[stat.ftype] + to_characters.join
+end
+
+def divide_characters(stat)
+  # mode.to_s(8)の戻り値の文字数は可変なため、末尾から指定しています。
+  stat.mode.to_s(8).slice(-3..-1).chars.each_with_index.to_h do |char, idx|
+    case idx
+    when 0 then [:user, char]
+    when 1 then [:group, char]
+    when 2 then [:other, char]
+    end
+  end
+end
+
+def to_special_permission(stat, section, char)
+  end_char_permisson = PERMISSIONS[char].slice(-1)
```

Comment:
「最後の文字の認可」と読めてしまうので、permission_end_char の方がよさそう。
ただ、寿命が短い変数なので、

```suggestion
  end_char = PERMISSIONS[char][-1]
```

でもいいと思います。（sliceも`[]`の方が短いのでいいかも。）

Reply:
`end_char`使わせていただきました。
`slice`メソッドの使用に関しては、`[-n]`の記述がrubocopで引っかかってしまうので使用しています🙏